### PR TITLE
Add debugger support on nuttx-smt32f4 target

### PR DESCRIPTION
**Pull request by robertsipka**
Using jerry_parse_named_resource to parse script and construct an EcmaScript function.
The file name will also passed to this function which is used by the debugger to find
the source code.
Run the constructed EcmaScript function instead of using the simple jerry runner.

### CODE REVIEW

File: targets/nuttx-stm32f4/jerry_main.c [hunk 1]
```diff
@@ -51,126 +52,64 @@ print_help (char *name)
 } /* print_help */
 
 /**
- * Read source files.
+ * Read source code into buffer.
  *
- * @return concatenated source files
+ * Returned value must be freed with jmem_heap_free_block if it's not NULL.
```
Comment (zherczeg) [inline targets/nuttx-stm32f4/jerry_main.c]: @return missing

File: targets/nuttx-stm32f4/jerry_main.c [hunk 2]
```diff
@@ -51,126 +52,64 @@ print_help (char *name)
...
  */
-static char*
-read_sources (const char *script_file_names[],
-              int files_count,
-              size_t *out_source_size_p)
+static const uint8_t *
+read_file (const char *file_name,
+           size_t *out_size_p)
```
Comment (zherczeg) [inline targets/nuttx-stm32f4/jerry_main.c]: Comments are needed for these arguments.

File: targets/nuttx-stm32f4/jerry_main.c [hunk 3]
```diff
@@ -51,126 +52,66 @@ print_help (char *name)
...
-      jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Failed to ftell script_len(%ld)\n", script_len);
-      fclose (file);
-      break;
-    }
-
-    total_length += (size_t) script_len;
-
-    fclose (file);
-    file = NULL;
+    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Error: failed to open file: %s\n", file_name);
```
Comment (zherczeg) [inline targets/nuttx-stm32f4/jerry_main.c]: cannot open file

File: targets/nuttx-stm32f4/jerry_main.c [hunk 4]
```diff
@@ -51,126 +52,66 @@ print_help (char *name)
...
+    return NULL;
   }
 
-  if (total_length <= 0)
+  int fseek_status = fseek (file, 0, SEEK_END);
+  if (fseek_status != 0)
   {
-    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "There's nothing to read\n");
+    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Failed to fseek fseek_status(%d)\n", fseek_status);
```
Comment (zherczeg) [inline targets/nuttx-stm32f4/jerry_main.c]: Failed to seek (error: %d)

File: targets/nuttx-stm32f4/jerry_main.c [hunk 5]
```diff
@@ -51,126 +52,66 @@ print_help (char *name)
...
     return NULL;
   }
 
-  source_buffer = (char*) malloc (total_length);
-  if (source_buffer == NULL)
+  long script_len = ftell (file);
+  if (script_len < 0)
   {
-    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Out of memory error\n");
+    jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Failed to ftell script_len(%ld)\n", script_len);
```
Comment (zherczeg) [inline targets/nuttx-stm32f4/jerry_main.c]: Failed to get the file size (error: %d)

File: targets/nuttx-stm32f4/jerry_main.c [hunk 6]
```diff
@@ -51,126 +52,66 @@ print_help (char *name)
...
-    long script_len = ftell (file);
-    if (script_len < 0)
-    {
-      jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Failed to ftell script_len(%ld)\n", script_len);
-      break;
-    }
+  rewind (file);
 
-    rewind (file);
+  uint8_t *buffer = jmem_heap_alloc_block (script_len);
```
Comment (zherczeg) [inline targets/nuttx-stm32f4/jerry_main.c]: This variant cannot return with NULL. It is a good idea to use the other here.

Comment (robertsipka) [inline targets/nuttx-stm32f4/jerry_main.c]: I'm not sure that I understand you correctly, but the jmem_heap_alloc_block can return with NULL if there is not enough memory. What do you mean 'other'?

Comment (robertsipka) [inline targets/nuttx-stm32f4/jerry_main.c]: Oh I missed it, you suggest to use `jmem_heap_alloc_block_internal` right?

Comment (zherczeg) [inline targets/nuttx-stm32f4/jerry_main.c]: void *jmem_heap_alloc_block (const size_t size);
    void *jmem_heap_alloc_block_null_on_error (const size_t size);

The jmem_heap_alloc_block crashes on error.

Comment (robertsipka) [inline targets/nuttx-stm32f4/jerry_main.c]: Ohh, I missed it, thanks! I've updated the PR.

File: targets/nuttx-stm32f4/jerry_main.c [hunk 7]
```diff
@@ -51,126 +52,66 @@ print_help (char *name)
...
+ * @return NULL, if read or allocation has failed
+ *         pointer to the allocated memory block, otherwise
  */
-static char*
-read_sources (const char *script_file_names[],
-              int files_count,
-              size_t *out_source_size_p)
+static const uint8_t *
+read_file (const char *file_name, /**< source code */
+           size_t *out_size_p) /**< number of bytes successfully read from source */
```
Comment (LaszloLango) [inline targets/nuttx-stm32f4/jerry_main.c]: `/**< [out] number of bytes successfully read from source */`

**Pull request merged**